### PR TITLE
src(buildpack): add python

### PR DESCRIPTION
**Pull request by lance**
This commit adds a python buildpack and simple invoker. This is awork in progress. Existing issues:

- Extremely slow on first build (>10 minutes) subsequent builds are really quick
- Attempt to install python 3.8 on UBI images results in 3.6?
- The installed pip is very old
- ~No user dependencies capability yet~
- ~Only CloudEvents functions are supported at the moment~
- ~Still needs a production runtime - at the moment, it's development mode only~

### CODE REVIEW

**Comment by zroubalik** [comment]
>     * Extremely slow on first build (>10 minutes)

@lance any idea why?

**Comment by lance** [comment]
@zroubalik really no idea why it's so slow. It happens when pulling in dependencies for flask, etc. If you can try a build locally and let me know if you see the same that would be great.

**Comment by lance** [comment]
@zroubalik it should be running for CE now

File: buildpacks/python/runtime/server.py [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+# create an endpoint at localhost:3000
+
+@app.route("/", methods=["POST"])
+def handler():
+  # create a CloudEvent
+  event = from_http(request.headers, request.get_data())
+  return func.main(event)
+
+if __name__ == "__main__":
+  app.run(port=8080)
```
Comment (matejvasek) [inline buildpacks/python/runtime/server.py:24]: @lance Does this start debug only or production ready server?

Comment (lance) [reply buildpacks/python/runtime/server.py:24]: Currently just debug. A production server is still on the todo list.

Comment (matejvasek) [reply buildpacks/python/runtime/server.py:24]: > Still needs a production runtime - at the moment, it's development mode only

Oh now I noticed.

**Comment by lance** [comment]
@matejvasek @zroubalik ptal - I think this might be ready to land. The  first three issues noted in the PR description are... :shrug:

**Comment by lance** [comment]
Considering creating and publishing a python package for the server so that we can support local testing and running on localhost similar to the Node.js buildpack.

**Comment by lance** [comment]
> Considering creating and publishing a python package for the server so that we can support local testing and running on localhost similar to the Node.js buildpack.

https://github.com/boson-project/parliament

Sample usage:

```
$ cat func.py
from cloudevents.http import CloudEvent, to_binary

# ---------------------------------------
# Function template
# context is a dictionary with the Flask request object and 
# and a cloud_event key for incoming events
# ---------------------------------------
def main(context):
    print(f"Method: {context['request'].method}")
    attributes = {
      "type": "com.example.fn",
      "source": "https://example.com/fn"
    }
    data = { "message": "Howdy!" }
    event = CloudEvent(attributes, data)
    headers, body = to_binary(event)
    return body, 200, headers
$ pip install parliament-functions
$ parliament .
```

**Comment by matejvasek** [comment]
@lance 
I got:
```
2021/03/29 17:04:36.506665 INFO:   ===> BUILDING
2021/03/29 17:04:36.506696 DEBUG:  Running the builder on OS linux with:
2021/03/29 17:04:36.506701 DEBUG:  Container Settings:
2021/03/29 17:04:36.506705 DEBUG:    Args: /cnb/lifecycle/builder
2021/03/29 17:04:36.506708 DEBUG:    System Envs: CNB_PLATFORM_API=0.4
2021/03/29 17:04:36.506712 DEBUG:    Image: pack.local/builder/696e6c706d6c71716b6e:latest
2021/03/29 17:04:36.506715 DEBUG:    User: 
2021/03/29 17:04:36.506725 DEBUG:    Labels: map[author:pack]
2021/03/29 17:04:36.506727 DEBUG:  Host Settings:
2021/03/29 17:04:36.506731 DEBUG:    Binds: pack-layers-qxdrlskzuf:/layers pack-app-nvzharrrsw:/workspace
2021/03/29 17:04:36.506735 DEBUG:    Network Mode: host
[builder] ---> Python Function Buildpack
[builder] /cnb/buildpacks/dev.boson.python/0.0.1/lib/build.sh: line 13: syntax error in conditional expression: unexpected token `;'
[builder] ERROR: failed to build: exit status 2
Error: executing lifecycle. This may be the result of using an untrusted builder: failed with status code: 145
```

I don't know why, `lib/build.sh` looks good to me.

**Comment by matejvasek** [comment]
```
shellcheck buildpacks/python/lib/build.sh                                                                                                                      ST 4   lance/add-python 

In buildpacks/python/lib/build.sh line 13:
  if [[ ! -f "${layer_dir}/environments/functions/bin/flask"]]; then
  ^-- SC1009: The mentioned syntax error was in this if expression.
     ^-- SC1073: Couldn't parse this test expression. Fix to allow more checks.
             ^-- SC1019: Expected this to be an argument to the unary condition.
                                                              ^-- SC1020: You need a space before the ]].
                                                              ^-- SC1072: Missing space before ]. Fix any mentioned problems and try again.

For more information:
  https://www.shellcheck.net/wiki/SC1020 -- You need a space before the ]].
  https://www.shellcheck.net/wiki/SC1019 -- Expected this to be an argument t...
  https://www.shellcheck.net/wiki/SC1072 -- Missing space before ]. Fix any m...
```

**Comment by matejvasek** [comment]
Hmm  maybe I checked out wrong remote for this.

**Comment by lance** [comment]
Strange... I definitely ran this BP over and over again last week.

**Comment by lance** [comment]
> Is there an easy way to throttle the number of revisions created?

Oh, sorry. I have not updated that. You should pull from the `lance/add-python` branch on this repo.

**Comment by matejvasek** [comment]
@lance  It's my fault I wrongfully assumed it's from your repo, the PR is OK, I think.

**Comment by matejvasek** [comment]
Yup it works from `git@github.com:boson-project/buildpacks.gi`.

**Comment by matejvasek** [comment]
@lance the python Function seems to react to `SIGINT` but it doesn't react to `SIGTERM` I think it should.

**Comment by matejvasek** [comment]
Also I noticed that the python Function runs under user id `1000`, whereas other functions (e.g. golang) run under uid of local user. I don't know if its good or bad or neither; or if it's deliberate.

**Comment by zroubalik** [comment]
> Also I noticed that the python Function runs under user id `1000`, whereas other functions (e.g. golang) run under uid of local user. I don't know if its good or bad or neither; or if it's deliberate.

Good catch, we should probably unify this.

**Review by zroubalik — APPROVED** [review]

**Comment by lance** [comment]
> Also I noticed that the python Function runs under user id 1000, whereas other functions (e.g. golang) run under uid of local user. I don't know if its good or bad or neither; or if it's deliberate.

Hmm this is confusing. Shouldn't they all run as user 1000?

https://github.com/boson-project/buildpacks/blob/10100d5462ac2706622c6c1a7cb0e7051810c9f3/stacks/ubi8/base/Dockerfile#L3-L12

**Comment by matejvasek** [comment]
> > Also I noticed that the python Function runs under user id 1000, whereas other functions (e.g. golang) run under uid of local user. I don't know if its good or bad or neither; or if it's deliberate.
> 
> Hmm this is confusing. Shouldn't they all run as user 1000?
> 
> https://github.com/boson-project/buildpacks/blob/10100d5462ac2706622c6c1a7cb0e7051810c9f3/stacks/ubi8/base/Dockerfile#L3-L12

Yes it should. Let me double check that. Maybe I used different docker daemons in different console windows. I sometimes use `podman` and sometimes `dockerd` might be caused by that.

**Comment by lance** [comment]
> the python Function seems to react to `SIGINT` but it doesn't react to `SIGTERM` I think it should.

I am a little stumped by this. The ENTRYPOINT command is 

```
source $environments_layer/functions/bin/activate && cd $app_layer && parliament .
```

If I run the equivalent commands locally and send `SIGTERM` to the process, it shuts down gracefully.

```
❯ parliament .
fish: “parliament .” terminated by signal SIGTERM (Polite quit request)
```

Do you think it would be OK if we don't let this block landing the PR, and instead open an issue for it? I did confirm that, for whatever reason, `SIGTERM` does not terminate the process when running the container.

**Comment by matejvasek** [comment]
@lance  With regard to UID -- it was indeed caused by using `podman` vs. `docker`.

**Comment by matejvasek** [comment]
>Do you think it would be OK if we don't let this block landing the PR, and instead open an issue for it? I did confirm that, for whatever reason, SIGTERM does not terminate the process when running the container.

:+1: I think that's OK. It's not any critical it will only prolong ksvc shutdown I think.

I can confirm, local run with sigterm work, but it doesn't work in container.

**Comment by lance** [comment]
Added an issue for `SIGTERM` here https://github.com/boson-project/buildpacks/issues/66

**Comment by matejvasek** [comment]
LGTM

**Comment by matejvasek** [comment]
Just one thing. I tried local development using venv. I created env in the project directory.
Then I got:
```
$f run  
ERROR: failed to launch: modify env: add layer env: apply env files from dir '/layers/dev.boson.python/app/env': read /layers/dev.boson.python/app/env/lib64: is a directory
```
I think that we somehow should let user know he is not supposed create `env` in the project directory.

**Review by matejvasek — APPROVED** [review]

**Comment by lance** [comment]
> I think that we somehow should let user know he is not supposed create env in the project directory.

@matejvasek do you know if that is standard practice for Python devs? When I was making all of this, I set up an environment in ~/src/environments and of course did not see this issue.

**Comment by matejvasek** [comment]
> > I think that we somehow should let user know he is not supposed create env in the project directory.
> 
> @matejvasek do you know if that is standard practice for Python devs? When I was making all of this, I set up an environment in ~/src/environments and of course did not see this issue.

Don't know, jetbrains IDE did that for me, but now I see it's creating `venv` (not `env`) by default. So I guess it's OK. Would be good to have seasoned python programmer to review this.

**Comment by lance** [comment]
`venv` is the standard naming that I have seen used as I worked on this.

**Comment by matejvasek** [comment]
> `venv` is the standard naming that I have seen used as I worked on this.
Then it's all right I think.

**Pull request merged**